### DISCUSSION TITLE
App Studio - Application Wide Extension Capability

### DISCUSSION THREAD

**Discussion by ernst86913** [directus/directus#18233] [Feature Requests]
### Summary

Currently the App Studio provides no way to run base extensions which would affect the whole application.

Extendibility is limited to specific screen components:  interfaces, displays, layouts, panels, operations. But those screen components are only rendered as a component in certain places and do not provide a way to inject an extension the affects the holistic application.

Custom Modules could be used to at least run code on module level - especially when using "custom views" instead of the standard "private view". Unfortunately any logic on that module level is only applicable when the custom module is opened... E.g. a pinia store for loading custom data which is needed for several custom modules would need to be initialized for every module instead once. Also custom routes which such apply for the whole application cannot be injected (remark: don't confuse those custom routes with API level routes).

On top, for unauthenticated user modules are not loaded at all which makes it impossible to eg have a custom login / registration component

FAZIT:  It would be great to have a way to hook application wide extensions into the Directus App Studio... 

### Basic Example

Examples that such "app base extensions" could be used for:

- Register additional App Studio application routes / pages (e.g. User Registration page,...)   (remark: don't confuse those routes with API routes)
- Override existing App Studio routes to adapt layout / functionality of existing pages
- Register a separate PINA store to fetch data one time centrally which can be used for any custom modules
- Re-route a user to a different page under certain conditions. A use case for this would be that the user needs to fill in some form data before the user is allowed to use directus application... a logic would check whether this has been completed already - if not the user is routed to a form page.

### Motivation

This would heavily improve extendability of the Directus Studio Application. The effort to implement such a hook should be relatively small compared to the benefits it provides.

### Detailed Design

I suggest to be able to load custom code somewhere in the main.ts file nearby the following code:

	registerDirectives(app);
	registerComponents(app);
	registerViews(app);
	await loadExtensions();
	registerExtensions(app);
	app.mount('#app');

### Requirements List

Be able to load custom "app base extension" with custom code which can affect the base functionality of the App Studio which for example includes routes and store.

### Must Have:
- Flexibility to load any custom code as part of the main Directus Studio application directly on root level, eg run addRoute / removeRoute, modify  router onBeforeEach and onAfterEach, run defineStore, hydrate a store.

### Won't Have:
- Will not replace/effect any existing App Studio extensions

### Drawbacks

Should potentially only be enabled on "custom installations". On Directus cloud installations I am not sure and might need to be disabled to not affect other tenants.

### Alternatives

Do nothing - means existing limitations of App extendability prevail. Users are forced to FORK the directus app to implement such changes which is combersome for anyone involved. Such forks are usually also bad from a community perspective as good ideas and good code gets lost in forks instead of finding its way back to Directus code base.

### Adoption Strategy

Not a breaking change... it opens up the door to much more flexible use cases of the Directus Studio application.

### Unresolved Questions

How and whether this can be used for Directus Cloud

**Comment by ernst86913** [comment]
Any thoughts?

**Reply by rijkvanzanten** [reply]
I can't really give any thoughts, as the important details are missing. The goal of the form template is to make sure that enough detailed information is available for both the team and the community to understand and give proper feedback, but that's currently impossible

**Reply by ernst86913** [reply]
@rijkvanzanten I tried to be more precise - please let me know if the reasoning / use cases are now better understandable - thx.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
